Source/DafnyStandardLibraries/src/Std/Parsers/String/StringParsers.dfy

```diff
@@ -0,0 +1,408 @@
+  }
+
+  opaque function Space(): (p: Parser<char>)
+    // A parser that tests if the current char is a whitespace including newline
+  {
+    CharTest(c => c in " \t\r\n", "space")
+  }
+
+  opaque function WS(): (p: Parser<string>)
+  {
+    ZeroOrMore(Space())
+  }
+
+  opaque function Digit(): (p: Parser<char>)
+    // A parser that tests if the current char is a digit and returns it
+  {
+    CharTest(c => c in "0123456789", "digit")
+  }
+
+  opaque function DigitNumber(): (p: Parser<nat>)
+    // A parser that returns the current char as a number if it is one
+  {
+    Map(Digit(), (c: char) =>
+          var d := DigitToInt(c);
+          var n: nat := if d >= 0 then d else 0;
+          n
+    )
+  }
+
+  opaque function Nat(): (p: Parser<nat>)
+    // A parser that parses a natural number
+  {
+    Bind(DigitNumber(),
+         (result: nat) =>
+           Rep(DigitNumber(),
+               (previous: nat, c: nat) =>
+                 var r: nat := previous * 10 + c; r,
+               result
+           )
+    )
+  }
+
+  opaque function Int(): (p: Parser<int>)
+    // A parser that parses a integer, possibly negative
+  {
+    Bind(Maybe(Char('-')),
+         (minusSign: Option<char>) =>
+           Map<nat, int>(Nat(), (result: nat) => if minusSign.Some? then 0-result else result))
+  }
+
+  opaque function String(expected: string): (p: Parser<string>)
+    // A parser that succeeds only if the input starts with the given string
+    // Its failure is recoverable, so it's possible to branch to something else
+  {
+    (input: Input) =>
+      if |expected| <= A.Length(input) && A.Slice(input, 0, |expected|).ToString() == expected then
+        ParseSuccess(expected, A.Drop(input, |expected|))
+      else ParseFailure(Recoverable, FailureData("expected '"+expected+"'", input, Option.None))
+  }
+
+  // ########################
+  // Error handling utilities
+  // ########################
+
+  function repeat_(str: string, n: nat): (r: string)
+    // Repeats the given string n times
+    ensures |r| == |str| * n
+  {
+    if n == 0 then ""
+    else str + repeat_(str, n-1)
+  }
+
+  datatype CodeLocation = CodeLocation(lineNumber: nat, colNumber: nat, lineStr: string)
+  datatype ExtractLineMutableState = ExtractLineMutableState(input: string, pos: nat, startLinePos: nat, i: nat, lineNumber: nat, colNumber: nat)
+
+  opaque ghost function ExtractLineColSpecAux(vars: ExtractLineMutableState): (res: ExtractLineMutableState)
+    requires 0 <= vars.startLinePos <= vars.i <= |vars.input|
+    decreases |vars.input| - vars.i
+    ensures 0 <= res.startLinePos <= res.i <= |res.input|
+    ensures !(res.i < |res.input| && res.i != res.pos) // Negation of the while guard
+  {
+    if vars.i < |vars.input| && vars.i != vars.pos then
+      var colNumber := vars.colNumber + 1;
+      if vars.input[vars.i] == '\r' && vars.i + 1 < |vars.input| && vars.input[vars.i+1] == '\n' then
+        //var startLinePos := i + 1; // ORIGINAL BUG: vars.i + 1
```

Comment:
Recording a time Dafny caught a bug? :)

Reply:
Yep, but the line has been changed so much that this is no longer relevant. Removing.